connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java

```diff
@@ -0,0 +1,386 @@
+    }
+
+    private List<TopicPartition> findTopicPartitions()
+            throws InterruptedException, ExecutionException {
+        Set<String> topics = listTopics(sourceAdminClient).stream()
+            .filter(this::shouldReplicateTopic)
+            .collect(Collectors.toSet());
+        return describeTopics(topics).stream()
+            .flatMap(MirrorSourceConnector::expandTopicDescription)
+            .collect(Collectors.toList());
+    }
+
+    private void refreshTopicPartitions()
+            throws InterruptedException, ExecutionException {
+        List<TopicPartition> topicPartitions = findTopicPartitions();
+        Set<TopicPartition> newTopicPartitions = new HashSet<>();
+        newTopicPartitions.addAll(topicPartitions);
+        newTopicPartitions.removeAll(knownTopicPartitions);
+        Set<TopicPartition> deadTopicPartitions = new HashSet<>();
+        deadTopicPartitions.addAll(knownTopicPartitions);
+        deadTopicPartitions.removeAll(topicPartitions);
+        if (!newTopicPartitions.isEmpty() || !deadTopicPartitions.isEmpty()) {
+            log.info("Found {} topic-partitions on {}. {} are new. {} were removed. Previously had {}.",
+                    topicPartitions.size(), sourceAndTarget.source(), newTopicPartitions.size(), 
+                    deadTopicPartitions.size(), knownTopicPartitions.size());
+            log.trace("Found new topic-partitions: {}", newTopicPartitions);
+            knownTopicPartitions = topicPartitions;
+            knownTargetTopics = findExistingTargetTopics(); 
+            createTopicPartitions();
+            context.requestTaskReconfiguration();
+        }
+    }
+
+    private void loadTopicPartitions()
+            throws InterruptedException, ExecutionException {
+        knownTopicPartitions = findTopicPartitions();
+        knownTargetTopics = findExistingTargetTopics(); 
+    }
+
+    private void refreshKnownTargetTopics()
+            throws InterruptedException, ExecutionException {
+        knownTargetTopics = findExistingTargetTopics();
+    }
+
+    private Set<String> findExistingTargetTopics()
+            throws InterruptedException, ExecutionException {
+        return listTopics(targetAdminClient).stream()
+            .filter(x -> sourceAndTarget.source().equals(replicationPolicy.topicSource(x)))
+            .collect(Collectors.toSet());
+    }
+
+    private Set<String> topicsBeingReplicated() {
+        return knownTopicPartitions.stream()
+            .map(x -> x.topic())
+            .distinct()
+            .filter(x -> knownTargetTopics.contains(formatRemoteTopic(x)))
+            .collect(Collectors.toSet());
+    }
+
+    private void syncTopicAcls()
+            throws InterruptedException, ExecutionException {
+        List<AclBinding> bindings = listTopicAclBindings().stream()
+            .filter(x -> x.pattern().resourceType() == ResourceType.TOPIC)
+            .filter(x -> x.pattern().patternType() == PatternType.LITERAL)
```

Comment:
Should we replicate prefix ACLs too?

Reply:
That would be neat, but there are landmines here. We don't want a prefix pattern to accidentally apply to topics on another cluster, so we can't just copy prefixes across. We could maybe expand a prefix "foo" on cluster "primary" to an equivalent "primary.foo" prefix, but that would only work if ReplicationPolicy used something like the default topic renaming convention -- a custom naming convention like "topic-DC1" would not work with prefixes this way.

We could maybe find all prefix patterns and replicate them as explicit literal patterns. So a prefix of "foo" turns into literals "primary.foo-1", "primary.foo-bar", etc. But even that is a little dangerous, as it is not easily reversible. For example, what happens when an upstream prefix ACL is deleted?

If there is a safe way to do this, I haven't thought of one yet. We should stick with literals only for now.